src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -0,0 +1,520 @@
+        comp->fgSetStmtSeq(info.stmt);
+    }
+
+    comp->fgMorphBlockStmt(lowerBndBb, lowerBndBb->lastStmt() DEBUGARG("Morph lowerBnd"));
+    comp->fgMorphBlockStmt(upperBndBb, upperBndBb->lastStmt() DEBUGARG("Morph upperBnd"));
+    if (lowerBndBb->lastStmt() != nullptr)
+    {
+        // lowerBndBb might be converted into no-op by fgMorphBlockStmt(lowerBndBb)
+        // it happens when we emit BBJ_COND(0 >= 0) fake block (for simplicity)
+        comp->gtUpdateStmtSideEffects(lowerBndBb->lastStmt());
+    }
+    comp->gtUpdateStmtSideEffects(upperBndBb->lastStmt());
+
+    // All blocks must be in the same EH region
+    assert(BasicBlock::sameEHRegion(prevBb, lowerBndBb));
+    assert(BasicBlock::sameEHRegion(prevBb, upperBndBb));
+    assert(BasicBlock::sameEHRegion(prevBb, fastpathBb));
+    assert(BasicBlock::sameEHRegion(prevBb, fallbackBb));
+    assert(BasicBlock::sameEHRegion(prevBb, lastBb));
+
+    return fastpathBb;
+}
+
+// A visitor to record all the bounds checks in a statement in the execution order
+class BoundsChecksVisitor final : public GenTreeVisitor<BoundsChecksVisitor>
+{
+    Statement*                      m_stmt;
+    ArrayStack<BoundCheckLocation>* m_boundsChks;
+
+public:
+    enum
+    {
+        DoPostOrder       = true,
+        UseExecutionOrder = true
+    };
+
+    BoundsChecksVisitor(Compiler* compiler, Statement* stmt, ArrayStack<BoundCheckLocation>* bndChkLocations)
+        : GenTreeVisitor(compiler)
+        , m_stmt(stmt)
+        , m_boundsChks(bndChkLocations)
+    {
+    }
+
+    fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTree* node = *use;
+        if (node->OperIs(GT_BOUNDS_CHECK))
+        {
+            m_boundsChks->Push(BoundCheckLocation(m_stmt, node->AsBoundsChk(), user));
+        }
+        return fgWalkResult::WALK_CONTINUE;
+    }
+};
+
+// -----------------------------------------------------------------------------
+// optRangeCheckCloning: The main entry point for the range check cloning phase.
+//    This phase scans all the blocks in the method and groups the bounds checks
+//    in each block by the "Base Index and Length" pairs (VNs). Then it picks up
+//    the largest group and clones the block to have fast and slow paths in order
+//    to optimize the bounds checks in the fast path.
+//    See the overview at the top of the file and the comments in the optRangeCheckCloning_DoClone
+//    function for more details.
+//
+// Return Value:
+//    The status of the phase after the transformation.
+//
+PhaseStatus Compiler::optRangeCheckCloning()
+{
+    if (!doesMethodHaveBoundsChecks())
+    {
+        JITDUMP("Current method has no bounds checks\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    if (lvaHaveManyLocals(0.75))
+    {
+        // If we're close to running out of locals, we should be conservative
+        JITDUMP("Too many locals - bail out.\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
```

Comment:
Why? Did you see any cases here?

I think it would be preferable to avoid checks like these if we can since it does not take anything about references or hotness/coldness into account. All that stuff is taken into account by the code that picks which locals to track.

Reply:
I've seen impact from this in my initial version I think, for this one - no diffs, will remove